referenceqvm/qvm_stabilizer.py

```diff
@@ -0,0 +1,479 @@
+        self.post()
+
+        # return HALTED (i.e. program_counter is end of program)
+        return self.program_counter == len(self.program)
+
+    def pre(self):
+        """
+        Pre-transition hook - use for noisy evolution models. Unimplemented for now.
+        """
+        pass
+
+    def post(self):
+        """
+        Post-transition hook - use for noisy evolution models. Unimplemented for now.
+        """
+        pass
+
+    def _transition(self, instruction):
+        """
+        Implements a transition on the generator matrix representing the stabilizers
+
+        :param Gate instruction: QuilAction gate to be implemented
+        """
+        if isinstance(instruction, Measurement):
+            # mutates classical memory!  I should change this...
+            self._apply_measurement(instruction)
+            self.program_counter += 1
+
+        elif isinstance(instruction, Gate):
+            # apply Gate or DefGate
+            if instruction.name == 'H':
+                self._apply_hadamard(instruction)
+            elif instruction.name == 'S':
+                self._apply_phase(instruction)
+            elif instruction.name == 'CNOT':
+                self._apply_cnot(instruction)
+            elif instruction.name == 'I':
+                pass
+            else:
+                raise TypeError("We checked for correct gate types previously" +
+                                " so the impossible has happened!")
+
+            self.program_counter += 1
+
+        elif isinstance(instruction, Jump):
+            # unconditional Jump; go directly to Label
+            self.program_counter = self.find_label(instruction.target)
+
+        elif isinstance(instruction, JumpTarget):
+            # Label; pass straight over
+            self.program_counter += 1
+
+    def run(self, pyquil_program, classical_addresses=None, trials=1):
+        """
+        Run pyquil program and return the results
+
+        Loads and checks program if all gates are within the stabilizer set.
+        Then executes program.  If measurements are requested then the measured
+        results are returned
+
+        :param Program pyquil_program: a pyquil Program containing only
+                                       CNOT-H-S-MEASUREMENT operations
+        :param classical_addresses: classical addresses to return
+        :param trials: number of times to repeat the execution of the program
+        :return: list of lists of classical memory after each run
+        """
+        self.load_program(pyquil_program)
+
+        if classical_addresses is None:
+            classical_addresses = get_classical_addresses_from_program(pyquil_program)
+
+        results = []
+        for _ in range(trials):
+            # set up stabilizers
+            self.tableau = self._n_qubit_tableau(self.num_qubits)
+            self.kernel()
+            results.append(list(map(int, self.classical_memory[classical_addresses])))
```

Comment:
`[int(b) for b in self.classical_memory[classical_addresses]]`

Reply:
👍